playground/backend/api.js

```diff
@@ -56,7 +69,82 @@ async function doFetch(url) {
       'Referer': 'https://amp.dev/playground',
     },
   });
+
+  if (!response.ok) {
+    throw new Error(`Request to ${fetchUrl} could not complete successfully.`);
+  }
+
+  const contentType = response.headers.get('content-type');
+  if (!contentType.includes('text/html')) {
+    throw new Error(`${fetchUrl} is no HTML document.`);
+  }
+
   return response.text();
 }
 
+const limits = new LRU({
+  maxAge: RATE_LIMIT_TIME_FRAME,
+});
+
+/**
+ * Verifies that the host to a certain URL has not been called
+ * more than DEFAULT_HOST_RATE_LIMIT
+ * @param  {URL} fetchUrl
+ * @return {Boolean}
+ */
+function exceedsRateLimit(fetchUrl) {
+  const host = fetchUrl.host;
+
+  // Requests to amp.dev are not affected by rate limiting
+  if (config.hostNames.has(host)) {
+    return false;
+  }
+
+  const count = (limits.peek(host) || 0) + 1;
+  if (count > HOST_RATE_LIMIT) {
+    return true;
+  }
+
+  limits.set(host, count);
+  limits.prune();
+  return false;
+}
+
+// eslint-disable-next-line new-cap
+const api = express.Router();
+api.get('/fetch', async (request, response) => {
+  let fetchUrl = request.query.url;
+  if (!fetchUrl) {
+    response.status(400).send('No URL provided.');
+    return;
+  }
+
+  fetchUrl = url.parse(fetchUrl);
+  if (!fetchUrl.protocol || !fetchUrl.host) {
+    response.status(400).send(`${fetchUrl} is not a valid URL.`);
+    return;
+  }
+
+  // Verify that this URL is currently allowed to be fetched
+  // and is not rate-limited
+  if (exceedsRateLimit(fetchUrl)) {
+    response
+      .status(429)
+      .send(
+        `${fetchUrl.host} has been requested too many times. ` +
+          `Please wait a few seconds and then try again.`
```

Comment:
@sebastianbenz, how transparent do we want to be towards the user? Rather just say something like "Request failed try again later"?

Reply:
this sounds good to me